### PR TITLE
ddl: handle cluster state retrieval failure

### LINKED ISSUE #53808: tidb restart with error log “[FATAL] [job_table.go:366] ["dispatch loop get cluster state failed, it should not happen, please try restart TiDB"]”
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、run sysbench mussel 
2、inject pd leader io delay 500ms

### 2. What did you expect to see? (Required)
tidb is normal

### 3. What did you see instead (Required)
tidb restart with error log “[FATAL] [job_table.go:366] ["dispatch loop get cluster state failed, it should not happen, please try restart TiDB"]”

`[2024/06/03 12:42:37.820 +08:00] [FATAL] [job_table.go:366] ["dispatch loop get cluster state failed, it should not happen, please try restart TiDB"] [category=ddl] [error="context canceled"] [stack="github.com/pingcap/tidb/pkg/ddl.(*jobScheduler).startDispatchLoop\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/ddl/job_table.go:366\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:171"]`
### 4. What is your TiDB version? (Required)

./tidb-server -V
 Release Version: v8.2.0-alpha
Edition: Community
Git Commit Hash: 8483dc538be95234dde6d6c31d1bb3248488440e
Git Branch: heads/refs/tags/v8.2.0-alpha
UTC Build Time: 2024-06-02 11:45:42
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2024-06-03T12:32:12.769+0800

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #53808

Problem Summary:

When TiDB is just elected as the DDL owner and `startDispatchLoop` is not started, there maybe a "context canceled" error when  owner retired for some reasons(like PD connection issue):

```
[2024/06/03 12:42:37.820 +08:00] [INFO] [manager.go:232] ["retire owner"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager 4e189321-5fd4-48fa-ad33-a8a92ed10cef"]
[2024/06/03 12:42:37.820 +08:00] [INFO] [syncer.go:485] ["schema version sync loop interrupted, retrying..."] [category=ddl]
[2024/06/03 12:42:37.820 +08:00] [WARN] [job_table.go:424] ["get global state failed"] [category=ddl] [error="context canceled"]
[2024/06/03 12:42:37.820 +08:00] [FATAL] [job_table.go:366] ["dispatch loop get cluster state failed, it should not happen, please try restart TiDB"] [category=ddl] [error="context canceled"] [stack="github.com/pingcap/tidb/pkg/ddl.(*jobScheduler).startDispatchLoop\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/ddl/job_table.go:366\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:171"]
```

### What changed and how does it work?

Since `startDispatchLoop` is not only called at TiDB cluster initialization, we should not os.Exit(1).

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/job_table.go [hunk 1]
```diff
@@ -354,7 +354,8 @@ func (d *ddl) startLocalWorkerLoop() {
 func (s *jobScheduler) startDispatchLoop() {
```
Comment (D3Hunter) [inline pkg/ddl/job_table.go:354]: we should wrap startDispatchLoop inside another loop to avoid transient network error，we can do it later

**Review by D3Hunter — APPROVED** [review]

File: pkg/ddl/job_table.go (hunk 1 above)
Comment (lance6716) [reply pkg/ddl/job_table.go:354]: After this PR, if the scheduler runs without the `startDispatchLoop`, does it have side effect to the cluster?

Comment (tangenta) [reply pkg/ddl/job_table.go:354]: All DDLs will be blocked until the cluster elects a new DDL owner.

Comment (lance6716) [reply pkg/ddl/job_table.go:354]: I think we'd better wrap a retry loop for `startDispatchLoop` in this PR. Just afraid the owner is still on this TiDB node and refresh lease succeed.

**Review by lance6716 — APPROVED** [review]

**Comment by lance6716** [comment]
@tangenta you can approve yourself

**Comment by tangenta** [comment]
> @tangenta you can approve yourself

Author cannot approve his own PR..

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by D3Hunter** [comment]
we should keep retry as long as we're still the owner, like what `SyncJobSchemaVerLoop` does

**Comment by tangenta** [comment]
> we should keep retry as long as we're still the owner, like what `SyncJobSchemaVerLoop` does

https://github.com/pingcap/tidb/pull/53816/commits/3fd88a1aa6d68d9c27763726a820be7b1805332c

File: pkg/ddl/job_table.go [hunk 2]
```diff
@@ -351,10 +351,33 @@ func (d *ddl) startLocalWorkerLoop() {
...
+		select {
+		case <-s.schCtx.Done():
+			logutil.DDLLogger().Info("startDispatchLoop quit due to context done")
+			return
+		case <-time.After(retryInterval):
+		}
+	}
+}
+
+func (s *jobScheduler) startDispatchLoop() error {
```
Comment (D3Hunter) [inline pkg/ddl/job_table.go:377]:
```suggestion
func (s *jobScheduler) startDispatch() error {
```

File: pkg/ddl/job_table.go [hunk 3]
```diff
@@ -351,10 +351,33 @@ func (d *ddl) startLocalWorkerLoop() {
 	}
 }
 
-func (s *jobScheduler) startDispatchLoop() {
+func (s *jobScheduler) startDispatchLoopWithRetry() {
```
Comment (D3Hunter) [inline pkg/ddl/job_table.go:354]:
```suggestion
func (s *jobScheduler) startDispatchLoop() {
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/ddl/job_table.go [hunk 4]
```diff
@@ -393,7 +416,7 @@ func (s *jobScheduler) startDispatchLoop() {
 				continue
 			}
 		case <-s.schCtx.Done():
-			return
+			return nil
```
Comment (D3Hunter) [inline pkg/ddl/job_table.go:419]:
```suggestion
			return s.schCtx.Err()
```

File: pkg/ddl/job_table.go [hunk 5]
```diff
@@ -351,10 +351,33 @@ func (d *ddl) startLocalWorkerLoop() {
...
 }
 
-func (s *jobScheduler) startDispatchLoop() {
+func (s *jobScheduler) startDispatchLoopWithRetry() {
+	const retryInterval = 3 * time.Second
+	for {
+		err := s.startDispatchLoop()
+		if err == nil {
+			return
+		}
```
Comment (D3Hunter) [inline pkg/ddl/job_table.go:360]:
```suggestion
```

**Pull request merged**